plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java

```diff
@@ -1154,21 +1181,27 @@ private ColumnMapping jsonColumnMapping()
         return ColumnMapping.sliceMapping(
                 jsonType,
                 (resultSet, columnIndex) -> jsonParse(utf8Slice(resultSet.getString(columnIndex))),
-                typedVarcharWriteFunction("json"),
+                typedVarcharWriteFunction("json", false),
                 DISABLE_PUSHDOWN);
     }
 
-    private static ColumnMapping typedVarcharColumnMapping(String jdbcTypeName)
+    private ColumnMapping typedVarcharColumnMapping(String jdbcTypeName, boolean withCollate)
     {
         return ColumnMapping.sliceMapping(
                 VARCHAR,
                 (resultSet, columnIndex) -> utf8Slice(resultSet.getString(columnIndex)),
-                typedVarcharWriteFunction(jdbcTypeName));
+                typedVarcharWriteFunction(jdbcTypeName, stringPushdownWithCollate && withCollate),
+                stringPushdownWithCollate && withCollate ? FULL_PUSHDOWN : POSTGRESQL_STRING_PUSHDOWN_WITHOUT_COLLATE);
```

Comment:
Can we refactor to have different type-mappings installed/used depending on whether `isStringPushdownWithCollateEnabled` is true or false?

That way one of the impl will be the old "simple" mapping and a new impl will provide the mapping to be used when collation needs to be specified?

Reply:
Hmm, I see. Because I don't have enough time rest of this week and probably next week, I will suspend this pull request for a while and resume working when I can find another spaire time.

Reply:
@hashhar What does `type-mappings` exactly mean here?

Reply:
I meant that the `toColumnMapping` and `toWriteMapping` methods can return different `ColumnMapping`/`WriteMapping`s depending on what the config is set to.

e.g. pseudo-java

```java
public WriteMapping toWriteMapping(ConnectorSession session, Type type)
{
    if (type instanceof VarcharType) {
        VarcharType varcharType = (VarcharType) type;
        String dataType;
        if (varcharType.isUnbounded()) {
            dataType = "varchar";
        }
        else {
            dataType = "varchar(" + varcharType.getBoundedLength() + ")";
        }
        
        if (isStringPushdownWithCollateEnabled) {
            return WriteMapping.sliceMapping(dataType, varcharWriteFunctionWithCollate());
        }

        return WriteMapping.sliceMapping(dataType, varcharWriteFunction());
     }
}
```

That way the actual `varcharWriteFunction` won't have too many conditionals and can be focused on doing just one thing (which it's name describes).

Reply:
Thanks for the explanation. Keeping many downstream objects simple and primitive by gathering complexity to a few upstream objects sounds a good trade!

Reply:
@hashhar Refactored. Do you think https://github.com/trinodb/trino/pull/9746/commits/1dc464865bc430fe7cc008b20b4e77a3903ba4e4 is better then the original code?